clients/src/main/java/org/apache/kafka/common/security/authenticator/SaslClientAuthenticator.java

```diff
@@ -229,6 +253,43 @@ public void authenticate() throws IOException {
         }
     }
 
+    @Override
+    public boolean supportsClientReauth() {
+        return true;
+    }
+
+    @Override
+    public void reauthenticate(ReauthenticationContext reauthenticationContext) throws IOException {
+        Authenticator previousAuthenticator = Objects.requireNonNull(reauthenticationContext).previousAuthenticator();
+        if (!(previousAuthenticator instanceof SaslClientAuthenticator))
+            throw new IllegalArgumentException(
+                    "Invalid previous authenticator in client-side re-authentication context: " + previousAuthenticator);
+        apiVersionsResponse = ((SaslClientAuthenticator) previousAuthenticator).apiVersionsResponse;
+        netInBuffer = reauthenticationContext.inProgressResponse();
+        reauthenticating = true;
+        interestedInWritingImmediatelyAfterReauthentication = transportLayer.selectionKey().isWritable();
```

Comment:
Do we need to track `interestedInWritingImmediatelyAfterReauthentication`? Aren't we starting reauthentication during write, so this would always be true?

Reply:
👍 Yes, it is always true -- I removed it.